README.rst
==============
file-processor
==============


.. image:: https://img.shields.io/pypi/v/file_processor.svg
        :target: https://pypi.python.org/pypi/file_processor

.. image:: https://img.shields.io/travis/amjadraza/file_processor.svg
        :target: https://travis-ci.com/amjadraza/file_processor

.. image:: https://readthedocs.org/projects/file-processor/badge/?version=latest
        :target: https://file-processor.readthedocs.io/en/latest/?version=latest
        :alt: Documentation Status




A project to processor files


* Free software: MIT license
* Documentation: https://file-processor.readthedocs.io.


Project Environment
-------------------

We create the project environment using below command.

``conda env create -f environment.yml -p ./venv``

Update the existing conda environment

``conda env update -f environment.yml -p ./venv``

Activate the environment

``conda activate ./venv``




Features
--------
This package is to read the text files using API.

Currently it supports below features

* An API to read the text files.
* Process the text file
* Notebooks with sanbox codes
* Package Documentation using Sphnix
* Other DevOps batteries

Run App using Python
---------------------
To develop/test, we can run the app using python with conda environment.

Test/run the using python with input arguments from project home directory

``python file_processor/main.py --file_name test_data/sample1.txt  --start_datetime 2000-01-01T17:25:49Z  --end_datetime 2000-01-06T06:27:36Z``

To set up the local development server using `FASTAPI` server

``uvicorn file_processor.API.main:app --reload``

Go to below link to test the App using `Docs`

```http://127.0.0.1:8000/docs```

To test the App using Postman, follow below link

```http://127.0.0.1:8000/process_file/```

Run Auto Tests
--------------------
We can run auto test of the functions for backend

use below commands

``python -m pytest tests/test_file_processor.py``


Run App using Docker
--------------------
This project includes `Dockerfile` to run the app in Docker container.

Build the docker container

``docker  build -t fileprocessor .``

Run the docker container

``docker run -d --name fp_app -p 80:80 fileprocessor``

Run the docker container using docker-compose

``docker-compose up``

Create Docs
------------
The documentation using Sphnix is already set it up.

Use below command to build from `.docs/` location

``make html``

TODO
----
* Streaming the text file
* Using Spark Streaming to read/process/store the files
* Load Balancer
* Host Docs
* Set up CI/CD


Credits
-------

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage


HISTORY.rst
=======
History
=======

0.1.0 (2021-11-19)
------------------

* First release on PyPI.


tests/__init__.py
"""Unit test package for file_processor."""


tests/test_file_processor.py
#!/usr/bin/env python

"""Tests for `file_processor` package."""
import sys
sys.path.append('.')
import pytest
from file_processor.utils.helpers import validate_iso8601_utc, filter_json_date

# @pytest.fixture
def test_utc_validator():

    assert validate_iso8601_utc('2000-01-06T06:27:36Z') is True

    assert validate_iso8601_utc('2000-01-54T06:27:36Z') is False

    assert validate_iso8601_utc('2000-01-06T06:27:36Z.06') is False


def test_file_processor():

    file_name = 'test_data/sample1.txt'
    start_datetime= '2000-01-01T17:25:49Z'
    end_datetime = '2000-01-06T06:27:36Z'

    try:
        response = filter_json_date(file_name, start_datetime, end_datetime)
        f = True
    except:
        f=False

    assert f is True




file_processor/utils/helpers.py
"""
.. module:: helpers
   :synopsis: helpers
.. moduleauthor:: MA Raza

This module contains helpers functions

Todo:
    * Add more helper functions
"""

import sys
sys.path.append('.')
import pandas as pd
from loguru import logger
import re

regex = r'^(-?(?:[1-9][0-9]*)?[0-9]{4})-(1[0-2]|0[1-9])-(3[01]|0[1-9]|[12][0-9])T(2[0-3]|[01][0-9]):([0-5][0-9]):([0-5][0-9]Z)?$'
match_iso8601 = re.compile(regex).match


def validate_iso8601_utc(str_val):
    """
    To check the valid ISO UTC format
    Args:
        str_val: UTC String format

    Returns: Bool

    """
    try:
        if match_iso8601( str_val ) is not None:
            return True
    except:
        pass
    return False


def filter_json_date(file_path, start_datetime, end_datetime):
    """
    This function reads the text file and filters based on given datetime ranges.
    After the filtering, df is deleted to free up the memory space.
    Args:
        file_path: Location of uploaded text file
        start_datetime: Start Datetime range in UTC format
        end_datetime: End Datetime range in UTC format

    Returns: Json Object

    """
    if validate_iso8601_utc(start_datetime) & validate_iso8601_utc(end_datetime):
        logger.info('Date Time formats are correct')
        logger.info('Loading the file')
        # df = pd.read_csv(StringIO(str(file_path.read(), 'utf-16')), encoding='utf-16')
        # logger.info(df.head())
        # df = pd.read_csv(file_path, sep=" ", header=None)
        # For larger file, we can use Chunks
        chunks = pd.read_csv(file_path, sep=" ", header=None, chunksize=1000000)
        df = pd.concat(chunks)

        # Using Multiprocessing to reading Chunks

        # logger.info(df.head())
        df.columns = ['eventTime', 'email', 'sessionId']
        logger.info('filtering the data')
        result = df[df['eventTime'].between(start_datetime, end_datetime)].to_json(orient='records', lines=False)
        # Delete the df to free up memory
        del df
        logger.info('The pandas df has been deleted to free up memory')
        return result
    else:
        logger.info('Input Date Time format is incorrect')
        return [{'message':'Input Date Time format is incorrect'}]


file_processor/API/endpoints/upload.py
import shutil
from typing import Any, List

from fastapi import APIRouter, Depends, HTTPException


from fastapi import FastAPI, UploadFile, File
from pathlib import Path
from tempfile import NamedTemporaryFile
import os
router = APIRouter()

FILES_DATA_DIR = '../../data_dir'
os.makedirs(os.path.join('.', FILES_DATA_DIR), exist_ok=True)

@router.post("/")
async def create_upload_files(id: int, files: List[UploadFile] = File(...)):
    id_dir = os.path.join(FILES_DATA_DIR, str(id))
    os.makedirs(id_dir, exist_ok=True)
    file_names = [file.filename for file in files]
    for file_name in file_names:
        destination = os.path.join(id_dir, file_name)
        try:
            with destination.open("wb") as buffer:
                shutil.copyfileobj(file_name, buffer)
        finally:
            file_name.close()

    return {"filenames": [file.filename for file in files]}


file_processor/API/main.py
"""
.. module:: main
   :synopsis: Main
.. moduleauthor:: MA Raza

This file is the entry point of API.

Example

    >>> uvicorn file_processor.API.main:app --reload

Todo:
    * Add more details/functionalities
    * Convert the Simple API into production ready API
"""
import sys
import os
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))
import os
import pandas as pd
from fastapi import FastAPI
import shutil
from typing import Any, List
from fastapi import APIRouter, Depends, HTTPException
from fastapi import FastAPI, UploadFile, File , Query
from pathlib import Path
from tempfile import NamedTemporaryFile
from typing import Optional
from file_processor.main import filter_json_date
from loguru import logger


FILES_DATA_DIR = 'data_dir/'
os.makedirs(os.path.join('.', FILES_DATA_DIR), exist_ok=True)

description = """
A Small App to render the filtered data from Text Files. 🚀

## Files

Upload text files with records in below format

`2020-12-04T11:14:23Z jane.doe@email.com 2f31eb2c-a735-4c91-a122-b3851bc87355`

## Date Time Ranges

Users will be able to provide:

* start_datetime: Start Datetime range in UTC format
* end_datetime: End Datetime range in UTC format
"""

# Define the App
app = FastAPI(openapi_url="/api/v1/openapi.json",
              title="File Processor",
              description=description,
              version="0.1.0",
              contact={
                "name": "MA Raza",
                "url": "https://github.com/amjadraza",
                "email": "amjadraza24@gmail.com",
                })

@app.post("/process_file/")
async def create_upload_file(
                             start_datetime: Optional[str]= Query('2000-01-01T17:25:49Z',
                                                  alias="Start DateTime",
                                                  title="Start datetime UTC Format",
                                                  ),
                             end_datetime: Optional[str]= Query('2000-01-06T06:27:36Z',
                                                  alias="End DataTie",
                                                  title="End datetime UTC Format",
                                                  ),
                             uploaded_file: UploadFile = File(...)):

    """
    API end point to run the filter. Users can upload the text file and datetime ranges t filter
    Args:
        start_datetime: Start Datetime range in UTC format
        end_datetime: End Datetime range in UTC format
        uploaded_file: text File

    Returns: Json format response
    """

    file_location = f"{FILES_DATA_DIR}/{uploaded_file.filename}"
    # Reading and saving the file
    with open(file_location, "wb+") as file_object:
        shutil.copyfileobj(uploaded_file.file, file_object)

    logger.info(f'File copied at: {file_location}')
    # processing the file
    response = filter_json_date(file_location, start_datetime, end_datetime)
    logger.info(response)

    return response
